### PR TITLE
infra/port: implement multicast l2 filtering

### LINKED ISSUE #27: implement unicast/multicast l2 filtering
A number of modules may require traffic from some unicast or multicast addresses that are filtered by hw when in non promisc / non allmulti mode.

Always running in promisc / allmulti is certainly possible, though it means receiving unwanted traffic.
But relying on promisc / allmulti may hide some network misconfiguration, or some other network equipment misbehaving, as grout would handle traffic that it should not in theory.

The request here is to add per iface lists of unicast and mcast addresses for which traffic is expected.

For a dpdk port backed interface, such lists will be associated with calls to rte_eth_dev_mac_addr_add, rte_eth_dev_mac_addr_del for unicast and rte_eth_dev_set_mc_addr_list. If adding a unicast/mcast address fails, then fallback would be to enable promisc / allmulti.

Associated with this, there is a need to track if enabling promisc/allmulti is a consequence of those unicast/mcast addresses or a request from the user (think of the case when the user enables promisc/allmulti via grcli, and later disables it).
The Linux kernel solution is to maintain a count of promisc / allmulti users (reflected by the "promiscuity" / "allmulti" counters one can find with "ip -s -d link").

### PR DESCRIPTION
**Pull request by christophefontaine**
- wrappers to call rte_eth_dev_mac_addr_[add|remove]
- Add API to filter multicast addresses on ports

resolves #27

### CODE REVIEW

**Comment by christophefontaine** [comment]
draft PR, to get some feedback on mcast filtering, before adding unicast filtering.

File: modules/infra/control/port.c [hunk 1]
```diff
@@ -457,6 +463,72 @@ static struct iface_type iface_type_port = {
...
+static int mcast_count(struct iface_info_port *p) {
+	int count = 0;
+	for (int i = 0; i < GR_PORT_MCAST_ADDR_MAX; i++) {
+		if (p->mcast_refcnt[i])
+			count++;
+	}
+	return count;
+}
+
+static void mcast_sort(struct iface_info_port *p) {
```
Comment (david-marchand) [inline modules/infra/control/port.c:475]: qsort() ?

File: modules/infra/control/port.c [hunk 2]
```diff
@@ -457,6 +463,72 @@ static struct iface_type iface_type_port = {
...
+		}
+	}
+}
+
+int gr_port_mcast_addr_add(struct iface_info_port *p, struct rte_ether_addr *mcast_addr) {
+	int i;
+	for (i = 0; i < GR_PORT_MCAST_ADDR_MAX; i++) {
+		if (rte_is_same_ether_addr(mcast_addr, &p->mcast[i])) {
+			LOG(DEBUG, "multicast address already filtered by port %d", p->port_id);
+			p->mcast_refcnt[i]++;
```
Comment (david-marchand) [inline modules/infra/control/port.c:497]: In which case do we have multiple modules requiring the same mcast address?

File: modules/infra/control/port.c [hunk 3]
```diff
@@ -409,6 +409,9 @@ static int iface_port_init(struct iface *iface, const void *api_info) {
 	if (port->devargs == NULL)
 		goto fail;
 
+	memset(port->mcast_refcnt, 0, sizeof(port->mcast_refcnt));
+	memset(port->mcast, 0, sizeof(port->mcast));
```
Comment (rjarry) [inline modules/infra/control/port.c:413]: I am not sure this is necessary since you are calling memcpy() just after.

**Review by rjarry — CHANGES_REQUESTED** [review]

File: modules/infra/control/port.c (hunk 2 above)
Comment (christophefontaine) [reply modules/infra/control/port.c]: Multiple protocols may use the same multicast mac address, or the same multicast IPs from 2 VRFs.

File: modules/infra/control/port.c (hunk 1 above)
Comment (christophefontaine) [reply modules/infra/control/port.c]: indeed !

**Review by rjarry — CHANGES_REQUESTED** [review]

File: modules/infra/control/port.c [hunk 4]
```diff
@@ -309,6 +309,28 @@ int iface_port_reconfig(
 			return errno_log(-ret, "rte_eth_macaddr_get");
 	}
 
+	if ((set_attrs & GR_PORT_ADD_MAC) && !eth_addr_is_zero(&api->mac)) {
```
Comment (rjarry) [inline modules/infra/control/port.c:312]: If the flag is set, should you actually check the value of the mac address? The other functions will fail if it is invalid.

File: modules/infra/control/port.c [hunk 5]
```diff
@@ -309,6 +309,28 @@ int iface_port_reconfig(
...
+		memcpy(&mac, &api->mac, sizeof(mac));
+		if (rte_is_multicast_ether_addr(&mac))
+			ret = gr_port_mcast_addr_add(p, &mac);
+		else
+			ret = gr_port_unicast_addr_add(p, &mac);
+		if (ret)
+			return errno_log(-ret, "add mac address");
+	}
+
+	if ((set_attrs & GR_PORT_REM_MAC) && !eth_addr_is_zero(&api->mac)) {
```
Comment (rjarry) [inline modules/infra/control/port.c:323]: If the flag is set, should you actually check the value of the mac address? The other functions will fail if it is invalid.

File: modules/infra/control/port.c [hunk 6]
```diff
@@ -406,8 +428,15 @@ static int iface_port_init(struct iface *iface, const void *api_info) {
 	if (port->devargs == NULL)
 		goto fail;
 
+	memset(port->multicast, 0, sizeof(port->multicast));
```
Comment (rjarry) [inline modules/infra/control/port.c:431]: Why is this needed?

File: modules/infra/control/port.c [hunk 7]
```diff
@@ -461,6 +508,162 @@ static struct iface_type iface_type_port = {
...
+		LOG(INFO,
+		    "Too many multicast addresses requested for %s",
+		    port_get_iface(p->port_id)->name);
+		goto enable_allmulti;
+	}
+
+	count = mcast_count(p);
+	for (i = 0; i < count; i++) {
+		memcpy(&mac[i], &p->multicast[i].mac, sizeof(struct rte_ether_addr));
+	}
```
Comment (rjarry) [inline modules/infra/control/port.c:551]: No need for braces.

Comment (rjarry) [reply modules/infra/control/port.c:551]:
```c
sizeof(mac[i])
```

File: modules/infra/control/port.c [hunk 8]
```diff
@@ -461,6 +508,162 @@ static struct iface_type iface_type_port = {
...
+		LOG(WARNING,
+		    "Enabling mc addr list: no space left on iface %s.",
+		    port_get_iface(p->port_id)->name);
+		p->flags |= GR_PORT_F_ALLMULTI_TOO_MANY_MCAST;
+	} else if (ret == -ENOTSUP) {
+		LOG(WARNING,
+		    "Enabling mc addr list not supported for iface %s",
+		    port_get_iface(p->port_id)->name);
+		p->flags |= GR_PORT_F_ALLMULTI_DEVICE_LIMIT;
+	}
```
Comment (rjarry) [inline modules/infra/control/port.c:567]: Minor style preferences:

* Do not use capital letters nor punctuation in log messages.
* Prefer shorter messages if possible.
* Maybe a switch case would be better here.

File: modules/infra/control/port.c [hunk 9]
```diff
@@ -461,6 +508,162 @@ static struct iface_type iface_type_port = {
...
+				return 0;
+			}
+		}
+	}
+
+	if (i == GR_PORT_NUM_MCAST_ADDR) {
+		LOG(INFO,
+		    "multicast addr wasn't allowed for iface %s",
+		    port_get_iface(p->port_id)->name);
+		return -ENODATA;
```
Comment (rjarry) [inline modules/infra/control/port.c:606]:
```c
return errno_log(ENOENT, "mcast_addr_del");
```

File: modules/infra/control/port.c [hunk 10]
```diff
@@ -461,6 +508,162 @@ static struct iface_type iface_type_port = {
...
+			return rte_eth_allmulticast_enable(p->port_id);
+		}
+	} else {
+		rte_eth_dev_set_mc_addr_list(p->port_id, NULL, 0);
+	}
+
+	iface = port_get_iface(p->port_id);
+	if (p->flags & ~GR_PORT_F_ALLMULTI_DEVICE_LIMIT && iface->flags & ~GR_IFACE_F_ALLMULTI) {
+		return rte_eth_allmulticast_disable(p->port_id);
+	}
```
Comment (rjarry) [inline modules/infra/control/port.c:631]: No braces please. And maybe this test can be shortened:

```c
if (p->flags & ~(GR_PORT_F_ALLMULTI_DEVICE_LIMIT | GR_IFACE_F_ALLMULTI))
```

File: modules/infra/control/port.c [hunk 11]
```diff
@@ -461,6 +508,162 @@ static struct iface_type iface_type_port = {
...
+
+int gr_port_unicast_addr_add(struct iface_info_port *p, struct rte_ether_addr *ucast_addr) {
+	int ret;
+
+	if (p->flags & GR_PORT_F_PROMISC_TOO_MANY_UNICAST
+	    || p->flags & GR_PORT_F_PROMISC_DEVICE_LIMIT)
+		return 0;
+
+	ret = rte_eth_dev_mac_addr_add(p->port_id, ucast_addr, 0);
+	switch (ret) {
```
Comment (rjarry) [inline modules/infra/control/port.c:644]: `ret` is used only here. Prefer shorter:

```c
switch (rte_eth_dev_mac_addr_add(p->port_id, ucast_addr, 0)) {
```

File: modules/infra/control/port.c [hunk 12]
```diff
@@ -461,6 +508,162 @@ static struct iface_type iface_type_port = {
...
+		return ret;
+		break;
+	}
+
+	return rte_eth_promiscuous_enable(p->port_id);
+}
+
+int gr_port_unicast_addr_del(struct iface_info_port *p, struct rte_ether_addr *ucast_addr) {
+	int ret = rte_eth_dev_mac_addr_remove(p->port_id, ucast_addr);
+	return ret;
```
Comment (rjarry) [inline modules/infra/control/port.c:664]:
```c
return rte_eth_dev_mac_addr_remove(p->port_id, ucast_addr);
```

**Review by rjarry — CHANGES_REQUESTED** [review]

File: modules/infra/cli/port.c [hunk 13]
```diff
@@ -25,6 +25,28 @@ static void port_show(const struct gr_api_client *, const struct gr_iface *iface
...
+		printf(" promisc (too many mac requested)");
+	if (port->flags & GR_PORT_F_PROMISC_DEVICE_LIMIT)
+		printf(" promisc (device)");
+	printf("\n");
+
+	for (int i = 0; i < GR_PORT_NUM_UCAST_ADDR; i++) {
+		if (!eth_addr_eq(&port->unicast[i], &port->mac)
+		    && !eth_addr_is_zero(&port->unicast[i]))
+			printf("additional mac: " ETH_ADDR_FMT "\n",
+			       ETH_BYTES_SPLIT(port->unicast[i].bytes));
```
[the remark this answers is not in the document]
Comment (christophefontaine) [reply modules/infra/cli/port.c:43]: The mac list always includes the default mac.

File: modules/infra/control/port.c (hunk 10 above)
Comment (christophefontaine) [reply modules/infra/control/port.c:631]: Can't be shortened as  port flags and iface flags are different.
